### ISSUE TITLE
Alternative to ⎕FIO ?

### ISSUE THREAD

**Issue by AtelesPaniscus** [exercism/gnu-apl#20]
I don't mean to tread on anyone's toes, take anyone out of their comfort zone or break any of Exercism's unwritten rules.

I am sorry to see that moving from GNU/APL 1.6 to 1.7 is giving problems.  I wonder if it is time to consider another way.

I use different languages for different purposes.  I try to play to their strengths.  Usually trying to do something the language permits but was not designed for gives scope for regret.

Take running a GNU/APL test case.  Currently I use:
```bash
apl -s -T *.tc
```
I confess I don't find that easy to remember so I would wrap it up in a one line shell script (and then embellish later as my needs dictate).

OK.  I use Linux and I don't think there is a version of Linux that does not have a shell (unless you count Android) and the same goes for MAC OS (unless you count iOS).  I can't see anyone using GNU/APL on a mobile because mobiles don't have a command line (which I why they have no shell).  Windows ?  Well, the GNU/APL web site says you can compile it under cygwin for Windows so even there there is a shell.   So I believe no one would have to install extra software just to run a one line script.

It appears to me that to run a test case, three files are concatenated together.  That is what the ⎕FIO stuff is doing.  The bare bones of a shell script that does this might look like:
```bash
echo "#!/usr/bin/env apl --script" > apl_test.tc
cat slug.apl test_harness.apl test_slug.apl >> apl_test.tc
```
GNU/APl generates a log and currently ⎕FIO is used to print this out.  That too could be replaced so a shell script that does the bare minimum might look like:
```bash
echo "#!/usr/bin/env apl --script" > slug.tc
cat slug.apl test_harness.apl test_slug.apl >> slug.tc

apl -s -T slug.tc

cat slug.tc.log
```
An added bonus is that folk attempting the exercise would not have to start their solution with "#!" line that clearly isn't APL (and isn't portable to other variants).

**Comment by marnen** [comment]
@AtelesPaniscus:

> I am sorry to see that moving from GNU/APL 1.6 to 1.7 is giving problems.

That is a grossly premature conclusion to jump to. The problems with 1.7, as far as we know now, are two:

1. 1.7 introduced a minor but breaking change in the semantics of `⎕FIO`. That's already fixed (by not calling `⎕FIO` directly) in #19.
2. 1.7 won't build on Ubuntu Trusty, which is the most recent OS available on Travis. That's easily fixed by having Travis use a Docker container with Xenial, and I'm about to change the configuration to do just that.

> I wonder if it is time to consider another way.

Not yet, I think. Talking about alternatives never hurts, of course.

> Currently I use: `apl -s -T *.tc`. I confess I don't find that easy to remember so I would wrap it up in a one line shell script (and then embellish later as my needs dictate).

Note that `apl -s -T *.tc` (and you don't need the `-s`) is already a one-line shell script... :)

> So I believe no one would have to install extra software just to run a one line script.

Not everyone has a POSIX environment, I think, and anyway, a shell script is not capable of manipulating the APL workspace itself in the way that an APL program can.

> It appears to me that to run a test case, three files are concatenated together.

Then  your impression is wrong. No files are concatenated.

> That is what the ⎕FIO stuff is doing. 

No it is not. Not at all. I agree that the numbered `⎕FIO` functions are confusing (which is part of the reason #19 exists!), but you can get a list of them quite easily by running `⎕FIO ''` in the interpreter. Please do so if you want to understand what test.apl is doing.

<blockquote>
a shell script that does the bare minimum might look like:

```bash
echo "#!/usr/bin/env apl --script" > slug.tc
cat slug.apl test_harness.apl test_slug.apl >> slug.tc

apl -s -T slug.tc

cat slug.tc.log
```
</blockquote>

I wouldn't expect that to work very well at all, because of the way that GNU APL's test runner works. Have you tried it?

> An added bonus is that folk attempting the exercise would not have to start their solution with "#!" line that clearly isn't APL (and isn't portable to other variants).

I don't like the shebang line either. However:

* It may not be necessary as often as I believe it is (my recommendation to include it was deliberately conservative, until I understand better when omitting it will break things)
* It is APL. GNU APL allows `#` as well as `⍝` for comments. (I agree, it's not _portable_ APL, and I wouldn't be sad to get rid of it.)

**Comment by AtelesPaniscus** [comment]
Oh, I'm sorry. I was not aware that the bare minimum shell script in my previous comment was not a working copy.  Did you try it ?

Below is a slightly more sophisticated version ... the ugly _sed_ line:

  - remove lines that feature )COPY just to make sure the concatenation isn't working by accident
  - remove the _#!_ lines from the input to demonstrate they are not needed
  - fiddles with the log file name so that _test∆show_log_ still works

It is there only to cope with there being two _.tc_ files and, if, were the script to be used, the line would disappear.

I haven not attempted to remove ⎕FIO 'on the fly'.  _test∆show_log_ is still required but does not use ⎕FIO (AFAIKT) and _test∆clear_log_ is not required since the script removes the log file before running GNU APL.
I suspect other instances of ⎕FIO would on longer be required but getting rid of ⎕FIO wasn't really what I wanted to demonstrate.

```bash
#!/bin/bash --posix                                                                                                                                                                                                                          
                                                                                                                                                                                                                                             
slug="${1-${PWD##*/}}";                                                                                                                                                                                                                      
test_hardness="../../test.apl";                                                                                                                                                                                                              
                                                                                                                                                                                                                                             
rm -f slug.tc;                                                                                                                                                                                                                               
echo "#!/usr/bin/env apl --script" > slug.tc;                                                                                                                                                                                                
cat "${slug}.apl" "${test_hardness}" "${slug}.tc" >> slug.tc;                                                                                                                                                                                
                                                                                                                                                                                                                                             
sed -e '/)COPY/d' -e '2,$s/^#!.*--script//' -e "s/${slug}.tc.log/slug.tc.log/" -i slug.tc;                                                                                                                                                   
                                                                                                                                                                                                                                             
rm -f slug.tc.log;                                                                                                                                                                                                                           
apl -s -T slug.tc;                                                                                                                                                                                                                           
                                                                                                                                                                                                                                             
# EOF                                                                                                                                                                                                                                        
```
                                                                                                                                                                                                                                                                                                                                                                                                                   
I understand the Exercism project has a strong preference for Go over bash.  The bash script is only intended to show that doing this kind of fiddling with files does not have to be done in APL.
